Command ranks

tastybento opened the pull request:
Added ranks and the ability to define which commands different ranks can run. Previously, some commands were only for island team leaders. Now, the server admin can set them to be available to sub-owners, or even members via a GUI.

tastybento committed:
Added command rank settings and extra ranks.

tastybento committed:
Merge remote-tracking branch 'origin/develop' into command_ranks

tastybento committed:
Fixed bug with locale files being saved from BentoBox into addon locales

tastybento committed:
Merge remote-tracking branch 'origin/develop' into command_ranks

tastybento committed:
Fixed test breaks.

tastybento committed:
Added team coop, uncoop, trust and untrust commands.

tastybento committed:
Added test classes for coop, uncoop, trust and untrust commands.

tastybento committed:
Fixed bug with loading add-on with no locales.

Fixed bug with no permission error in click listeners.

tastybento committed:
Added BentoBox reload command

Only reloads locales for now.

tastybento committed:
Enforced onlyConsole setting.

Removed extraneous test code method

Poslovitch commented on pom.xml:
```diff
@@ -5,7 +5,7 @@
 	<groupId>world.bentobox</groupId>
 	<artifactId>bentobox</artifactId>
-	<version>FC-0.92</version>
+	<version>FC-0.95</version>
```
Okay!

Poslovitch commented on src/main/java/world/bentobox/bentobox/Settings.java:
```diff
@@ -109,6 +110,10 @@
+    
+    @ConfigComment("Rank required to use a command. e.g., use the invite command. Default is owner rank is required.")
+    @ConfigEntry(path = "general.rank-command")
+    private Map<String, Integer> rankCommand = new HashMap<>();
```
I don't understand how this would work : can you provide some default examples in the HashMap you're instantiating?

tastybento replied:
Here's a copy from my config.yml:

```
  # Rank required to use a command. e.g., use the invite command. Default is owner rank is required.
  rank-command:
    /island team uncoop: 900
    /ai unban: 900
    /ai team kick: 500
    /ai ban: 500
    /ai team invite: 500
    /ai team untrust: 900
    /ai team coop: 900
    /island ban: 900
    /ai team trust: 900
    /island team coop: 900
    /island team untrust: 500
```

The key is unique because it's a fully-qualified command. The rank is the minimum rank value required, just like protection.

Poslovitch commented on src/main/java/world/bentobox/bentobox/api/commands/CompositeCommand.java:
```diff
@@ -821,4 +826,33 @@ protected boolean checkCooldown(User user, UUID targetUUID) {
+    /**
+     * This command can be configured for use by different ranks
+     */
+    public void setConfigurableRankCommand() {
```
I think it should be renamed to `useConfigurableRank`. And this method should have a boolean parameter.

tastybento replied:
This just comes from the variable (configurableRankCommand), which is true if the command is a configurable rank command. CongifurableRank would have a different meaning in my mind. No need for a boolean parameter as it's only used if it is going to be set to true.

Poslovitch commented on src/main/java/world/bentobox/bentobox/commands/BentoBoxCommand.java:
```diff
@@ -16,8 +16,10 @@ public BentoBoxCommand() {
 
     @Override
     public void setup() {
+        setPermission("bentobox.admin");
```
Only BentoBoxReloadCommand should need the permission. "version" and "about" are doing no harm.

tastybento replied:
ok

Poslovitch reviewed (dismissed):
The overall is good, however there are some details that need to be addressed before merging.

Poslovitch commented on src/main/java/world/bentobox/bentobox/api/commands/CompositeCommand.java:
```diff
@@ -51,6 +45,17 @@
+    /**
+     * True if the command is only for the console to use
+     */
+    private boolean onlyConsole = false;
```
I don't have a single clue about any command that might one day need it, but, well, let's keep it :D

tastybento replied:
I was thinking of some commands for BentoBox that would show a lot of text or stats about the addons. It's just in case though.

Poslovitch committed:
Brought back FC-0.92

I need to merge this pull request in order to continue working on BentoBox (some changes might cause conflicts). Therefore, I just put the version back to FC-0.92.
